problem-1

```diff
@@ -0,0 +1,36 @@
+//brute force 2nd sol
+//time : m*n
+//space : O(1)
+class Solution {
+    public int strStr(String haystack, String needle) {
+    
+      int i=0;
+        int n=haystack.length();
+        int m=needle.length();
+        if(m==0 ) return 0;
+      //int j=0;
+       while(i<haystack.length())
+       {
+
+           int j=0;
+          if(haystack.charAt(i)==needle.charAt(j))
+          {
+               int temp=i;
+              while(temp<n && j<m)
+              {
+                  if(haystack.charAt(temp)!=needle.charAt(j))
+                  {
+                     break;
+                  }
+                  temp++;
+                  j++; 
+              }
+              if(j==m)
+                  return i;
+             
+          }
+           i++;
+    }
+        return -1;
+}
+}
```

Comment:
The optimal solution for this problem is below
class Solution {
    public int strStr(String haystack, String needle) {
        int m = haystack.length();
        int n = needle.length();
        if(n == 0) return 0;
        int [] lps = lps(needle, n);
        int i = 0; int j = 0;
        while(i < m){
            if(haystack.charAt(i) == needle.charAt(j)){
                i++; j++;
            }
            if(j == n) {
                return i - j;
            } else if(i < m && j > 0 && haystack.charAt(i) != needle.charAt(j)){
                j = lps[j-1];
            } else if (i < m && j == 0 && haystack.charAt(i) != needle.charAt(j)){
                i++;
            }
        }
        return -1;
        
    }
    private int [] lps (String pattern, int n){
        int [] lps = new int [n];
        int i = 1;
        int j = 0;
        lps[0] = 0;
        while(i < n){
            // Case 1
            if(pattern.charAt(j) == pattern.charAt(i)){
                j++; lps[i] = j; i++;
            } else if (j > 0 && pattern.charAt(j) != pattern.charAt(i)){
                j = lps[j-1];
            } else if (j == 0 && pattern.charAt(j) != pattern.charAt(i)){
                lps[i] = 0;
                i++;
            }
        }
        return lps;
    }
}